### ISSUE TITLE
Poor react performance: epic

### ISSUE THREAD

**Issue by vic-topcoder** [appirio-tech/connect-app#1090]
We have a lot of bad components and we should pay big attention to creating a front-end performant app. That's the whole point we're using React after all! It is very very very very very important to optimize the code once a component is created, as it can cause huge degradation of front-end rendering and slows down the whole app.

Following the techniques outline here https://medium.freecodecamp.org/make-react-fast-again-tools-and-techniques-for-speeding-up-your-react-app-7ad39d3c1b82 one can clearly see that we do update almost all of our components on every click. An apparent issue is the user drop-down which is updated on every click (!) in the app, not only when the drop-down menu is visible (as expected to close it). THAT IS HUGE. Small things like that pile up. We can do better.

Install React Dev Tools and check "highlight updates", install "why-did-you-update" and you'll get a notification for unnecessary updates. You'll see a notification on each component change, and clearly we shouldn't update the whole page if I'm just checking a checkbox (in a checkbox group) or clicking around (dropdown) unless the dropdown is active. Video or the issues here: https://youtu.be/OgDsURgtZGs

Once we do an analysis of all pages, we can clearly identify components that should not update and we must create a list of all issues and fix them.

We should be careful of how we use our components, guys, and avoid trashing the performance.

**Comment by vic-topcoder** [comment]
To use "why-did-you-update" add the following in `index.jsx` right under the declarations

```javascript
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line no-unused-vars,react/no-deprecated
  let createClass = React.createClass;
  Object.defineProperty(React, 'createClass', {
    set: (nextCreateClass) => {
      createClass = nextCreateClass;
    },
  });
  // eslint-disable-next-line global-require
  const { whyDidYouUpdate } = require('why-did-you-update');
  whyDidYouUpdate(React);
}
```

**Comment by fnisen** [comment]
@vikasrohit, @gondzo who wants to take this one?

**Comment by gondzo** [comment]
I can take this one and handle it after this release. It would be a good idea to add these checks in the challenge reviews.

**Comment by vic-topcoder** [comment]
That works for me! It is important to verify all our components are not causing unnecessary renders. Thanks!

**Comment by fnisen** [comment]
I'll keep it as a P1 b/c it's important, but it's not part of the "Logged out project creation" project, so we don't need it for the release.

**Comment by vikasrohit** [comment]
Hey @vic-topcoder Thanks for the suggesting the tools and potential problems. I agree that there are few things that can be improved. I have even tried installing 'why-did-you-update` and in fact after installing that plugin, it made the whole application significantly slow and that is expected upto some extent for any performance tool.

I don't think the mentioned issues or similar to them, are creating very big performance lag anywhere in the app. I am not able to see major performance lags in the application. Yes, we can and we should optimize the way we are using React components, however, that optimization does not fall under basic usage of React. We have to first use defaults settings/configuration of any framework/library when we start using them and gradually optimize its usage by closely monitoring the areas where we can improve the usage. Same is with React. We tried to create dumb components if possible which are already optimized by React for performance. Parth and me also tried to optimize the form components which are most likely to cause frequent renders, but we always faced issues when we try to use PureComponent with them or tried to implement the shouldComponentUpdate on ourselves. This happened with every library that we tried for forms.

Coming back to the 'why-did-you-update', when I tried that I have also noticed that it logs every field of a JSON prop as an opportunity to reduce a re-render while actually there was only one or two re-renders e.g.
<img width="607" alt="screen shot 2017-08-28 at 5 02 49 pm" src="https://user-images.githubusercontent.com/2417632/29771694-d13a928e-8c12-11e7-8d71-412c83419c78.png">
In the above picture, we could not save 5 re-renders as said by the tool, because there was only one(I have checked that by putting a console log for every render) re-render of the FillProjectDetails component. So, it seems simply seeing the count of logs made by 'why-did-you-update' does not give the hint of how many unnecessary re-renders we are causing.
Further, I may be wrong,  but there are cases where we can't avoid re-renders e.g. if anything changes in project json, few wrapper components like CreateContainer, ProjectWizard, FillProjectDetails and ProjectBasicDetailsForm because they have to pass all such changes to child components.

Being said all these, I definitely agree that we are doing many re-renders which can be avoided. However, I would strongly suggest that we do that optimization only one by one because it can cause significant regressions if we optimize multiple components at once. To start with, it would be great if @vic-topcoder and @fnisen you guys can share some of the pain points in terms of performance from anywhere in the application. I can start debugging that and may be can find the root cause and optimize it.

**Comment by vic-topcoder** [comment]
More reading on the topic - how to avoid re-renders on components:
https://medium.freecodecamp.org/why-arrow-functions-and-bind-in-reacts-render-are-problematic-f1c08b060e36 and https://medium.freecodecamp.org/react-pattern-extract-child-components-to-avoid-binding-e3ad8310725e

I checked some of our code and there are examples when we do use super complex structure to create a component, so no abstraction was used to make sure we separate the logical structures into child components and further ensure performance. I think performance is an issue as we do reload almost all our content everywhere. This is not what I expect. The simplest example - when I navigate between pages in the projects list, the whole page is re-rendered. That's not needed. Anytime I go to a new page, we render everything. We use .SVG in 4-5 different ways. We rebuild 4-5 different implementations of the same component. All those are making the code unstable and cause A-performance issues and B-style regressions. All I urge is to take things under control and come up with a way that would fix the performance now, because there's not going to be a cycle dedicated to code refactoring. We cannot postpone architecture, it should always be included as part of a release.

**Comment by vikasrohit** [comment]
@vic-topcoder Agree that we can avoid renders at many places but there are places where we can't. Further react renders are not heavy. And the major benefit of using React is that unless any thing in DOM has changed, it won't update DOM. So even if we call render multiple times, it won't update the DOM if there is no change in the DOM. Please refer https://stackoverflow.com/a/35641329/365172
Accepting the need of optimization constructively, all I am trying to convey is that these optimizations are not so trivial(I have gone through many components again to check why there are so many renders, and found that there are one or another condition which requires the component to render. Of course there are many instances which can be improved but any most of those places we can rely on React's virtual DOM comparison for optimum performance) and falls under basic usage of React. These needs to be evaluated based on per instance basis. :)

**Comment by vic-topcoder** [comment]
As long as we know and monitor performance, and we don't create unnecessary redraws (as I mentioned before - we redraw the complete page when loading the list of projects, or apply any filters to the list, or search, and such things...), I'm good.

In reality there are several big components that need to be refactored (I guess) and that would speed up things, but we need an analysis on what are the usual culprits on issues, and keep a list of issues we want to fix. I do agree that randomly trying to fix every render is not optimal, nor should it be done. I just point towards the resources in the beginning as achieving good/smooth transitions and UI state is something we should always strive to achieve. 

Example: showing/hiding filters on manager view on the project list view. It is just turning the element on and off (removing it/adding it to the DOM) and there's no transition. It is a jagged experience that breaks the experience as the user doesn't perceive the change visually. I tried to remedy this by a simple CSS animation of the top bar container that pushes the whole list down. Flickering, jumping, reloading elements that are not changing - all that CAN be fixed through proper techniques in React. This is all I want to achieve. Hope that makes sense.

**Comment by vikasrohit** [comment]
Sorry, @vic-topcoder , if it seems like I am presenting counter statements, however, I am trying to present a rationale behind things why they are like the way they are and what are the problems in optimizing them. :)

**TL;DR** I full agree that we need to start doing optimizations as and when we see the opportunity. Below I am just trying to understand the exact pain point of examples you shared, so that I can jump on fixing them.

I didn't understand how we can prevent redraw of complete page when someone changes page. Specifically when something is changed in topbar (e.g. searching a particular project or apply any filter), it is required to re-render all components on the page. Without that how we would render the latest content to the DOM? Only opportunity I see here is that we can prevent top bar from being re-rendered when nothing is changed in filters, search or user dropdown menu. I think I am to do that, not sure how much performance that would increase and what side effects it can have. I am starting a separate git branch to do all these performance optimizations.

I am also unable to understand how removing/adding the filters section from DOM, is a related to best practices of React. There is nothing in React itself which can improve that so called flicker of hide/show of elements that are being removed/added to the DOM. It depends on the UX we want. As you also tried to fix it with CSS, I think that is the only solution for this use case.

**Comment by vikasrohit** [comment]
I have pushed few changes into feature/performance-tuning branch for optimizing the renders of project details form related components. It requires specific branch from react-components repo, which I am struggling now to add to shrinkwrap. Once that is done, you should be able to notice changes. Even without that specific branch of react-components, it contains more optimization changes that should be visible on project details form.
And while doing this, I also updated the SVGIconImage component to render svg as inline object instead of rendering it as image. I know this was pending from very long time. It is not easy to get good quality output from community. This change now allows converting few stateful components to functional components.

**Comment by vic-topcoder** [comment]
@vikasrohit - I am not saying we should fix everything that can be fixed, as there is a need for cost/benefit analysis here, as you explained above. I 100% agree we should optimize only things that make sense. What I think is the problem is our architectural approach to React and instead of breaking things into smaller components and handling them, we have a massive parent controller that pushes changes into the smaller components and then everything is redrawn. We're not working on the architecture optimization and keeping all our repos on the same level, nor we have a baseline that we measure against for major features going forward. Please consider the points I'm making more of a philosophical nature and drive for consistency on the codebase, rather than immediate need.

Thanks,
Vic

**Comment by vikasrohit** [comment]
@vic-topcoder Deployed those performance tuning changes to connect-test01.topcoder-dev.com.

**Comment by vikasrohit** [comment]
Just for reference, I am pasting few relevant links about formsy-react components being rendered on every other component change:

https://github.com/christianalfoni/formsy-react/issues/350
https://github.com/christianalfoni/formsy-react/issues/154
https://github.com/christianalfoni/formsy-react/issues/134

**Comment by vic-topcoder** [comment]
@mtwomey maybe this is bigger part of tech debt. Capture in Aha! and close if you see fit.

**Comment by vikasrohit** [comment]
Just for information, I have fixed few of the things here:

- Stopped re-render of other form fields when one of the fields in the form is updated
- Loaded SVG images as inline svg object
- Optimized ProjectWizard, TopbarContainer components to no re-render for every render of its parent

**Comment by vic-topcoder** [comment]
Can we get this merged into Dev? I did see better page performance when I tested.

**Comment by vikasrohit** [comment]
@vic-topcoder these are merged with the two tier changes. So, they would be there on production when we merged two tier changes to master.

**Comment by vikasrohit** [comment]
Just fyi, I have optimized code for Dropdown component to not handle outside click event when the dropdown is not open. This change is done for #1349 and yet not merged with dev so may not be visible on dev.

**Comment by vic-topcoder** [comment]
Closing the issue - it's too generic right now
